{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Chapter 03: Use of NS-Forest\n",
    "\n",
    "Ray LeClair \\<2025-01-08 Wed\\>\n",
    "\n",
    "## Objectives\n",
    "\n",
    "Since use of NS-Forest for discovery of minimum marker gene\n",
    "combinations for cell type identification from single-cell RNA\n",
    "sequencing data is the first processing step of the NLM Cell Knowledge\n",
    "Network pipeline, the objective of this document is to:\n",
    "\n",
    "-   Run NS-Forest on an example human lung cell CELLxGENE dataset\n",
    "\n",
    "### Background\n",
    "\n",
    "The NS-Forest repository contains a tutorial that walks through all\n",
    "aspects of using NS-Forest.\n",
    "\n",
    "The following sections describe various development environments\n",
    "\n",
    "See:\n",
    "\n",
    "-   [NS-Forest Tutorial](https://nsforest.readthedocs.io/en/latest/tutorial.html)\n",
    "-   [springbok-nlm-kn/README.md](https://github.com/ralatsdc/springbok-nlm-kn/blob/main/README.md)\n",
    "\n",
    "### Jupyter Notebook development environment\n",
    "\n",
    "Launch Jupyter Notebook from a terminal in which `.zshenv` has been\n",
    "sourced, and the virtual environment has been activated.\n",
    "\n",
    "### Emacs Org Mode development environment\n",
    "\n",
    "Launch Emacs from a terminal in which `.zshenv` has been sourced, then\n",
    "evaluate this code block to activate the virtual environment:\n",
    "\n",
    "``` commonlisp\n",
    "(pyvenv-activate \"../../.venv\")\n",
    "```\n",
    "\n",
    "## Run NS-Forest on an example human lung cell CELLxGENE dataset\n",
    "\n",
    "Following the tutorial, we write a function that runs NS-Forest on an\n",
    "example human lung cell CELLxGENE dataset. We assume the dataset has\n",
    "been previously identified and downloaded.\n",
    "\n",
    "See: [Chapter-01-CELLxGENE.ipynb](Chapter-01-CELLxGENE.ipynb)\n",
    "\n",
    "To begin, we import modules, and assign module scope variables:"
   ],
   "id": "8f0bcea8-2945-43db-9d2a-c6d503c579c6"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "results": "silent",
    "session": "shared",
    "tangle": "../py/NSForest.py"
   },
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import nsforest as ns\n",
    "from nsforest import nsforesting\n",
    "import scanpy as sc\n",
    "\n",
    "DATA_DIR = \"../data\"\n",
    "\n",
    "CELLXGENE_DIR = f\"{DATA_DIR}/cellxgene\"\n",
    "\n",
    "NSFOREST_DIR = f\"{DATA_DIR}/nsforest-2024-06-27\"\n",
    "TOTAL_COUNTS = 5000  # TODO: Select a more sensible value\n",
    "\n",
    "NCBI_CELL_DIR = f\"{DATA_DIR}/ncbi-cell\"\n"
   ],
   "id": "2cb77c9f-8cbe-4ad1-bb57-adb36af5ee24"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we write the function, noting:\n",
    "\n",
    "-   Some datasets have multiple annotations per sample\n",
    "    (ex. `broad_cell_type` and `granular_cell_type`). NSForest can be\n",
    "    run on multiple `cluster_header` values. Combining the parent and\n",
    "    child markers may improve classification results.\n",
    "\n",
    "-   `adata.var_names` must be unique. If there is a problem, usually it\n",
    "    can be solved by assigning `adata.var.index = adata.var[\"ensembl_id\"]`.\n",
    "\n",
    "-   Some datasets are too large and need to be downsampled to be run\n",
    "    through the pipeline. When downsampling, be sure to have all the\n",
    "    granular cluster annotations represented.\n",
    "\n",
    "-   Only run ns.pp.dendrogram() if there is no pre-defined dendrogram\n",
    "    order. This step can still be run with no effects, but the runtime\n",
    "    may increase."
   ],
   "id": "ee642d27-e931-4ca3-9745-74bf94e058d9"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "results": "silent",
    "session": "shared",
    "tangle": "../py/NSForest.py"
   },
   "outputs": [],
   "source": [
    "def run_nsforest_on_file(h5ad_filename, cluster_header=\"cell_type\"):\n",
    "    \"\"\"Run NSForest using the specified dataset filename, and\n",
    "    cluster_header.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    h5ad_filename : str\n",
    "       The dataset filename\n",
    "    cluster_header : str\n",
    "       The cluster header\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    None\n",
    "    \"\"\"\n",
    "    # Assign results filename and directory\n",
    "    pp_h5ad_filename = f\"pp_{h5ad_filename}\"\n",
    "    results_dirname = h5ad_filename.split(\".\")[0]\n",
    "    results_dirpath = f\"{NSFOREST_DIR}/{results_dirname}\"\n",
    "\n",
    "    # Run NSForest if results do not exist\n",
    "    if not os.path.exists(results_dirpath):\n",
    "        os.makedirs(results_dirpath)\n",
    "\n",
    "        print(f\"Loading unprocessed AnnData file: {h5ad_filename}\")\n",
    "        h5ad_filepath = f\"{CELLXGENE_DIR}/{h5ad_filename}\"\n",
    "        up_adata = sc.read_h5ad(h5ad_filepath)\n",
    "\n",
    "        # TODO: Check validity of downsampling\n",
    "        print(\"Calculating QC metrics\")\n",
    "        up_metrics = sc.pp.calculate_qc_metrics(up_adata)\n",
    "        if up_metrics[1][\"total_counts\"].sum() > TOTAL_COUNTS:\n",
    "            print(\"Downsampling unprocessed AnnData file\")\n",
    "            ds_adata = sc.pp.downsample_counts(\n",
    "                up_adata, total_counts=TOTAL_COUNTS, copy=True\n",
    "            )\n",
    "        else:\n",
    "            ds_adata = up_adata  # No need to copy\n",
    "\n",
    "        print(\"Generating scanpy dendrogram\")\n",
    "        # Dendrogram order is stored in\n",
    "        # `pp_adata.uns[\"dendrogram_cluster\"][\"categories_ordered\"]`\n",
    "        pp_adata = up_adata.copy()\n",
    "        pp_adata.obs[cluster_header] = pp_adata.obs[cluster_header].astype(str)\n",
    "        pp_adata.obs[cluster_header] = pp_adata.obs[cluster_header].astype(\"category\")\n",
    "        pp_adata = ns.pp.dendrogram(\n",
    "            pp_adata,\n",
    "            cluster_header,\n",
    "            save=True,\n",
    "            show=True,\n",
    "            output_folder=results_dirpath,\n",
    "            outputfilename_suffix=cluster_header,\n",
    "        )\n",
    "\n",
    "        print(\"Calculating cluster medians per gene\")\n",
    "        pp_adata = ns.pp.prep_medians(pp_adata, cluster_header)\n",
    "\n",
    "        print(\"Calculating binary scores per gene per cluster\")\n",
    "        pp_adata = ns.pp.prep_binary_scores(pp_adata, cluster_header)\n",
    "\n",
    "        pp_h5ad_filepath = f\"{results_dirpath}/{pp_h5ad_filename}\"\n",
    "        print(f\"Saving preprocessed AnnData file: {pp_h5ad_filepath}\")\n",
    "        pp_adata.write_h5ad(pp_h5ad_filepath)\n",
    "\n",
    "        print(f\"Running NSForest for preprocessed AnnData file: {pp_h5ad_filename}\")\n",
    "        results = nsforesting.NSForest(\n",
    "            pp_adata,\n",
    "            cluster_header,\n",
    "            output_folder=f\"{results_dirpath}/\",\n",
    "            outputfilename_prefix=cluster_header,\n",
    "        )\n",
    "\n",
    "        # Create dendrogram to plot\n",
    "        dendrogram = []  # custom dendrogram order\n",
    "        dendrogram = list(\n",
    "            pp_adata.uns[\"dendrogram_\" + cluster_header][\"categories_ordered\"]\n",
    "        )\n",
    "\n",
    "        # Create results to plot\n",
    "        to_plot = results.copy()\n",
    "        to_plot[\"clusterName\"] = to_plot[\"clusterName\"].astype(\"category\")\n",
    "        to_plot[\"clusterName\"] = to_plot[\"clusterName\"].cat.set_categories(dendrogram)\n",
    "        to_plot = to_plot.sort_values(\"clusterName\")\n",
    "        to_plot = to_plot.rename(columns={\"NSForest_markers\": \"markers\"})\n",
    "        to_plot.head()\n",
    "        markers_dict = dict(zip(to_plot[\"clusterName\"], to_plot[\"markers\"]))\n",
    "\n",
    "        print(\"Generating scanpy dotplot\")\n",
    "        ns.pl.dotplot(\n",
    "            pp_adata,\n",
    "            markers_dict,\n",
    "            cluster_header,\n",
    "            dendrogram=dendrogram,\n",
    "            save=True,\n",
    "            show=True,\n",
    "            output_folder=results_dirpath,\n",
    "            outputfilename_suffix=cluster_header,\n",
    "        )\n",
    "\n",
    "        print(\"Generating scanpy stacked violin plot\")\n",
    "        ns.pl.stackedviolin(\n",
    "            pp_adata,\n",
    "            markers_dict,\n",
    "            cluster_header,\n",
    "            dendrogram=dendrogram,\n",
    "            save=True,\n",
    "            show=True,\n",
    "            output_folder=results_dirpath,\n",
    "            outputfilename_suffix=cluster_header,\n",
    "        )\n",
    "\n",
    "        print(\"Generating scanpy matrix plot\")\n",
    "        ns.pl.matrixplot(\n",
    "            pp_adata,\n",
    "            markers_dict,\n",
    "            cluster_header,\n",
    "            dendrogram=dendrogram,\n",
    "            save=True,\n",
    "            show=True,\n",
    "            output_folder=results_dirname,\n",
    "            outputfilename_suffix=cluster_header,\n",
    "        )\n",
    "\n",
    "    else:\n",
    "        print(f\"Completed NSForest for preprocessed AnnData file: {pp_h5ad_filename}\")\n"
   ],
   "id": "14798a2b-5c41-4b32-9348-3b2f70552e08"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now call the function for an example CELLxGENE dataset using the\n",
    "default `cluster_header` of `\"cell_type\"`:"
   ],
   "id": "83770183-5d51-48f3-8668-ec04cd6dce74"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "results": "output",
    "session": "shared"
   },
   "outputs": [],
   "source": [
    "try:\n",
    "    h5ad_filename = \"6e00ccf7-0749-46ef-a999-dba785630d52.H5AD\"\n",
    "    run_nsforest_on_file(h5ad_filename, cluster_header=\"cell_type\")\n",
    "except Exception:\n",
    "    print_exc()\n"
   ],
   "id": "c8b7855a-608e-4209-8de5-17b1d857278c"
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, in Chapter 04 we investigate the use of OntoGPT for publication\n",
    "processing. In Chapter 05 we'll use the results produced by NS-Forest\n",
    "to populate an ArangoDB database graph.\n",
    "\n",
    "See:\n",
    "\n",
    "-   [Chapter-04-OntoGPT.ipynb](Chapter-04-OntoGPT.ipynb)\n",
    "-   [Chapter-05-ArangoDB.ipynb](Chapter-05-ArangoDB.ipynb)"
   ],
   "id": "3a0e2839-1fd4-464f-9c17-800e3d1243b3"
  }
 ],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {}
}
