{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "928d26c8-598e-487f-a1ac-4e6a0266a52c",
   "metadata": {},
   "source": [
    "# Chapter 01: Use of CELLxGENE\n",
    "\n",
    "Ray LeClair \\<2024-10-23 Wed\\>\n",
    "\n",
    "## Objectives\n",
    "\n",
    "Since CELLxGENE serves as an initiating data source for the NLM\n",
    "Knowledge Network pilot, the objectives of this document include\n",
    "demonstration of:\n",
    "\n",
    "-   Identification of CELLxGENE datasets for specified organisms, and\n",
    "    tissues\n",
    "\n",
    "-   Identification of publications corresponding to CELLxGENE datasets\n",
    "\n",
    "-   Determination of the dataset filename and dataset file download\n",
    "\n",
    "### Background\n",
    "\n",
    "All single-cell RNA data from Chan Zuckerberg (CZ) CELLxGENE Discover\n",
    "is accessed, queried, and analyzed using the CELLxGENE Discover\n",
    "Census. Using cell-based slicing and querying one can:\n",
    "\n",
    "-   Interact with the data through TileDB-SOMA\n",
    "\n",
    "-   Get slices in AnnData, Seurat, or SingleCellExperiment objects\n",
    "\n",
    "See: [CELLxGENE Discover Census](https://chanzuckerberg.github.io/cellxgene-census/)\n",
    "\n",
    "The following sections draw from CZ CELLxGENE tutorials, or a Chan\n",
    "Zuckerberg Initiative (CZI) repository, which demonstrate how to:\n",
    "\n",
    "-   [Explore and query the Census in the context of a single tissue, lung](https://chanzuckerberg.github.io/cellxgene-census/notebooks/analysis_demo/comp_bio_explore_and_load_lung_data.html)\n",
    "\n",
    "-   [Query the expression data and cell/gene metadata from the Census, and load them into common in-memory Python objects](https://chanzuckerberg.github.io/cellxgene-census/notebooks/api_demo/census_query_extract.html)\n",
    "\n",
    "-   [Generate a citation string for all datasets contained in a Census slice](https://chanzuckerberg.github.io/cellxgene-census/notebooks/api_demo/census_citation_generation.html)\n",
    "\n",
    "-   [Fetch full metadata for a Dataset](https://github.com/chanzuckerberg/single-cell-curation/blob/0c77179d2e794846861f8109c037b723507959cb/notebooks/curation_api/python_raw/get_dataset.ipynb)\n",
    "\n",
    "The following sections describe various development environments\n",
    "\n",
    "See: [springbok-nlm-kn/README.md](https://github.com/ralatsdc/springbok-nlm-kn/blob/main/README.md)\n",
    "\n",
    "### Jupyter Notebook development environment\n",
    "\n",
    "Launch Jupyter Notebook from a terminal in which `.zshenv` has been\n",
    "sourced, and the virtual environment has been activated.\n",
    "\n",
    "### Emacs Org Mode development environment\n",
    "\n",
    "Launch Emacs from a terminal in which `.zshenv` has been sourced, then\n",
    "evaluate this code block to activate the virtual environment:\n",
    "\n",
    "``` commonlisp\n",
    "(pyvenv-activate \"../../.venv\")\n",
    "```\n",
    "\n",
    "## Identification of CELLxGENE datasets for specified organisms, and tissues\n",
    "\n",
    "Mostly following the first tutorial, we write a function that obtains\n",
    "all datasets, summary cell counts, gene metadata, and, by default,\n",
    "human and mouse lung, eye, and brain cell metadata from the CZ\n",
    "CELLxGENE Census as Pandas DataFrames. Anticipating a time consuming\n",
    "process, the first call of the function writes the DataFrames to\n",
    "`.parquet` files, then, on subsequent calls, it reads the `.parquet`\n",
    "files. In both cases, the resulting DataFrames are returned.\n",
    "\n",
    "To begin, we import modules, and assign module scope variables:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dc749322-bd73-477b-85ca-69201538f082",
   "metadata": {
    "results": "silent",
    "session": "shared",
    "tangle": "../py/CELLxGENE.py"
   },
   "outputs": [],
   "source": [
    "import logging\n",
    "import os\n",
    "import re\n",
    "import subprocess\n",
    "from time import sleep\n",
    "from traceback import print_exception\n",
    "\n",
    "import cellxgene_census\n",
    "\n",
    "from bs4 import BeautifulSoup\n",
    "import pandas as pd\n",
    "import requests\n",
    "\n",
    "DATA_DIR = \"../data\"\n",
    "\n",
    "CELLXGENE_DOMAIN_NAME = \"cellxgene.cziscience.com\"\n",
    "CELLXGENE_API_URL_BASE = f\"https://api.{CELLXGENE_DOMAIN_NAME}\"\n",
    "CELLXGENE_DIR = f\"{DATA_DIR}/cellxgene\"\n",
    "\n",
    "NLM_KN_DIR = f\"{DATA_DIR}/nlm-kn\"\n",
    "\n",
    "HTTPS_SLEEP = 1\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "444d8add-7551-4cff-9f21-fac6659aebcd",
   "metadata": {},
   "source": [
    "Next we write the function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b5c5d799-acb9-4754-9ae7-8096631cd402",
   "metadata": {
    "results": "silent",
    "session": "shared",
    "tangle": "../py/CELLxGENE.py"
   },
   "outputs": [],
   "source": [
    "def get_metadata_and_datasets(\n",
    "    organisms=[\"homo_sapiens\", \"mus_musculus\"], tissues=[\"lung\", \"eye\", \"brain\"]\n",
    "):\n",
    "    \"\"\"Use the CZ CELLxGENE Census to obtain all datasets, summary\n",
    "    cell counts, gene metadata, and, by default, human and mouse\n",
    "    lung, eye, and brain cell metadata, then write the resulting\n",
    "    Pandas DataFrames to parquet files, or, if the files exist, read\n",
    "    them.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    organisms : list(str)\n",
    "        List of organisms, default is [\"homo_sapiens\", \"mus_musculus\"]\n",
    "    tissues : list(str)\n",
    "        List of tissues, default is [\"lung\", \"eye\", \"brain\"]\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    datasets : pd.DataFrame\n",
    "        DataFrame containing dataset descriptions\n",
    "    counts : pd.DataFrame\n",
    "        DataFrame containing summary cell counts\n",
    "    var : pd.DataFrame\n",
    "        DataFrame containing gene metadata\n",
    "    obs : pd.DataFrame\n",
    "        DataFrame containing cell metadata\n",
    "    \"\"\"\n",
    "    # Create and write, or read DataFrames\n",
    "    datasets_parquet = f\"{NLM_KN_DIR}/datasets.parquet\"\n",
    "    counts_parquet = f\"{NLM_KN_DIR}/counts.parquet\"\n",
    "    var_parquet = f\"{NLM_KN_DIR}/var.parquet\"\n",
    "    obs_parquet = f\"{NLM_KN_DIR}/obs.parquet\"\n",
    "    if (\n",
    "        not os.path.exists(datasets_parquet)\n",
    "        or not os.path.exists(counts_parquet)\n",
    "        or not os.path.exists(var_parquet)\n",
    "        or not os.path.exists(obs_parquet)\n",
    "    ):\n",
    "        print(\"Opening soma\")\n",
    "        census = cellxgene_census.open_soma(census_version=\"latest\")\n",
    "\n",
    "        print(\"Collecting all datasets\")\n",
    "        datasets = census[\"census_info\"][\"datasets\"].read().concat().to_pandas()\n",
    "\n",
    "        print(\"Collecting summary cell counts\")\n",
    "        counts = (\n",
    "            census[\"census_info\"][\"summary_cell_counts\"].read().concat().to_pandas()\n",
    "        )\n",
    "\n",
    "        var = pd.DataFrame()\n",
    "        for organism in organisms:\n",
    "            print(f\"Collecting gene metadata for {organism}\")\n",
    "            var = pd.concat(\n",
    "                [\n",
    "                    var,\n",
    "                    cellxgene_census.get_var(\n",
    "                        census,\n",
    "                        organism,\n",
    "                    ),\n",
    "                ]\n",
    "            )\n",
    "\n",
    "        obs = pd.DataFrame()\n",
    "        for organism in organisms:\n",
    "            print(f\"Collecting cell metadata for {organism}: {tissues} tissue\")\n",
    "            obs_for_org = cellxgene_census.get_obs(\n",
    "                census,\n",
    "                organism,\n",
    "                value_filter=f\"tissue_general in {tissues} and is_primary_data == True\",\n",
    "            )\n",
    "            obs_for_org[\"organism\"] = organism\n",
    "            obs = pd.concat(\n",
    "                [\n",
    "                    obs,\n",
    "                    obs_for_org,\n",
    "                ]\n",
    "            )\n",
    "\n",
    "        print(\"Closing soma\")\n",
    "        census.close()\n",
    "\n",
    "        print(\"Writing datasets parquet\")\n",
    "        datasets.to_parquet(datasets_parquet)\n",
    "\n",
    "        print(\"Writing summary cell counts parquet\")\n",
    "        counts.to_parquet(counts_parquet)\n",
    "\n",
    "        print(\"Writing gene metadata parquet\")\n",
    "        var.to_parquet(var_parquet)\n",
    "\n",
    "        print(\"Writing cell metadata parquet\")\n",
    "        obs.to_parquet(obs_parquet)\n",
    "\n",
    "    else:\n",
    "\n",
    "        print(\"Reading datasets parquet\")\n",
    "        datasets = pd.read_parquet(datasets_parquet)\n",
    "\n",
    "        print(\"Reading summary cell counts parquet\")\n",
    "        counts = pd.read_parquet(counts_parquet)\n",
    "\n",
    "        print(\"Reading gene metadata parquet\")\n",
    "        var = pd.read_parquet(var_parquet)\n",
    "\n",
    "        print(\"Reading cell metadata parquet\")\n",
    "        obs = pd.read_parquet(obs_parquet)\n",
    "\n",
    "    return datasets, counts, var, obs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dca6c29e-118e-4beb-b32b-1df512d11dee",
   "metadata": {},
   "source": [
    "Then call it to obtain the human and mouse lung, eye, and brain cell\n",
    "metadata and datasets (using exception handling since accessing an\n",
    "external resource), and print the result:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "08195cc8-1333-425d-90bb-55574b1e7087",
   "metadata": {
    "results": "output",
    "session": "shared"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading datasets parquet\n",
      "Reading summary cell counts parquet\n",
      "Reading gene metadata parquet\n",
      "Reading cell metadata parquet\n",
      "datasets:\n",
      "\n",
      "columns: Index(['soma_joinid', 'citation', 'collection_id', 'collection_name',\n",
      "       'collection_doi', 'collection_doi_label', 'dataset_id',\n",
      "       'dataset_version_id', 'dataset_title', 'dataset_h5ad_path',\n",
      "       'dataset_total_cell_count'],\n",
      "      dtype='object')\n",
      "\n",
      "     soma_joinid                                           citation  \\\n",
      "0              0  Dataset Version: https://datasets.cellxgene.cz...   \n",
      "1              1  Publication: https://doi.org/10.1002/hep4.1854...   \n",
      "2              2  Publication: https://doi.org/10.1126/sciimmuno...   \n",
      "3              3  Publication: https://doi.org/10.1038/s41593-02...   \n",
      "4              4  Publication: https://doi.org/10.1038/s41467-02...   \n",
      "..           ...                                                ...   \n",
      "951          951  Publication: https://doi.org/10.1101/2023.11.0...   \n",
      "952          952  Publication: https://doi.org/10.1038/s41586-02...   \n",
      "953          953  Publication: https://doi.org/10.1038/s41586-02...   \n",
      "954          954  Publication: https://doi.org/10.1038/s41593-02...   \n",
      "955          955  Publication: https://doi.org/10.1038/s41593-02...   \n",
      "\n",
      "                            collection_id  \\\n",
      "0    a96133de-e951-4e2d-ace6-59db8b3bfb1d   \n",
      "1    44531dd9-1388-4416-a117-af0a99de2294   \n",
      "2    3a2af25b-2338-4266-aad3-aa8d07473f50   \n",
      "3    180bff9c-c8a5-4539-b13b-ddbc00d643e6   \n",
      "4    bf325905-5e8e-42e3-933d-9a9053e9af80   \n",
      "..                                    ...   \n",
      "951  4c6eaf5c-6d57-4c76-b1e9-60df8c655f1e   \n",
      "952  45d5d2c3-bc28-4814-aed6-0bb6f0e11c82   \n",
      "953  45d5d2c3-bc28-4814-aed6-0bb6f0e11c82   \n",
      "954  1ca90a2d-2943-483d-b678-b809bf464c30   \n",
      "955  1ca90a2d-2943-483d-b678-b809bf464c30   \n",
      "\n",
      "                                       collection_name  \\\n",
      "0    HTAN/HTAPP Broad - Spatio-molecular dissection...   \n",
      "1    Single-Cell, Single-Nucleus, and Spatial RNA S...   \n",
      "2    Single-cell analysis of human B cell maturatio...   \n",
      "3    Molecular characterization of selectively vuln...   \n",
      "4    Single-cell Atlas of common variable immunodef...   \n",
      "..                                                 ...   \n",
      "951              Single cell atlas of the human retina   \n",
      "952  A single-cell transcriptional timelapse of mou...   \n",
      "953  A single-cell transcriptional timelapse of mou...   \n",
      "954  SEA-AD: Seattle Alzheimer’s Disease Brain Cell...   \n",
      "955  SEA-AD: Seattle Alzheimer’s Disease Brain Cell...   \n",
      "\n",
      "                 collection_doi  \\\n",
      "0                                 \n",
      "1             10.1002/hep4.1854   \n",
      "2    10.1126/sciimmunol.abe6291   \n",
      "3    10.1038/s41593-020-00764-7   \n",
      "4    10.1038/s41467-022-29450-x   \n",
      "..                          ...   \n",
      "951   10.1101/2023.11.07.566105   \n",
      "952  10.1038/s41586-024-07069-w   \n",
      "953  10.1038/s41586-024-07069-w   \n",
      "954  10.1038/s41593-024-01774-5   \n",
      "955  10.1038/s41593-024-01774-5   \n",
      "\n",
      "                                collection_doi_label  \\\n",
      "0                                                      \n",
      "1    Andrews et al. (2022) Hepatology Communications   \n",
      "2                   King et al. (2021) Sci. Immunol.   \n",
      "3                    Leng et al. (2021) Nat Neurosci   \n",
      "4          Rodríguez-Ubreva et al. (2022) Nat Commun   \n",
      "..                                               ...   \n",
      "951                         Li et al. (2023) bioRxiv   \n",
      "952                         Qiu et al. (2024) Nature   \n",
      "953                         Qiu et al. (2024) Nature   \n",
      "954               Gabitto et al. (2024) Nat Neurosci   \n",
      "955               Gabitto et al. (2024) Nat Neurosci   \n",
      "\n",
      "                               dataset_id  \\\n",
      "0    d7476ae2-e320-4703-8304-da5c42627e71   \n",
      "1    0895c838-e550-48a3-a777-dbcd35d30272   \n",
      "2    00ff600e-6e2e-4d76-846f-0eec4f0ae417   \n",
      "3    bdacc907-7c26-419f-8808-969eab3ca2e8   \n",
      "4    a5d95a42-0137-496f-8a60-101e17f263c8   \n",
      "..                                    ...   \n",
      "951  d6505c89-c43d-4c28-8c4f-7351a5fd5528   \n",
      "952  dcfa2614-7ca7-4d82-814c-350626eccb26   \n",
      "953  dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3   \n",
      "954  c2876b1b-06d8-4d96-a56b-5304f815b99a   \n",
      "955  6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3   \n",
      "\n",
      "                       dataset_version_id  \\\n",
      "0    acaf4493-139d-4be8-b50f-f8e74fe54c44   \n",
      "1    17616a35-1b3d-4754-9553-8ddb71267d63   \n",
      "2    3fa969b4-25e8-4e0b-b2ec-86cd9ddd2ccb   \n",
      "3    017837df-b8be-4a4f-bc08-c5f14bd8815a   \n",
      "4    f223d2b0-d878-443b-8bef-f182fe1f22d1   \n",
      "..                                    ...   \n",
      "951  488d9376-5b2f-4557-93ed-8524903aa00a   \n",
      "952  3002a659-a1a9-4406-9976-99e658e1fbb5   \n",
      "953  3817734b-0f82-433b-8c38-55b214200fff   \n",
      "954  c32964d2-3339-441f-8e56-7177234c7876   \n",
      "955  d3427e8c-c55d-4d4e-b15b-1a8774cd3a4b   \n",
      "\n",
      "                                         dataset_title  \\\n",
      "0                         HTAPP-330-SMP-1082 scRNA-seq   \n",
      "1                         Healthy human liver: B cells   \n",
      "2                 Human tonsil nonlymphoid cells scRNA   \n",
      "3    Molecular characterization of selectively vuln...   \n",
      "4                     Steady-state B cells - scRNA-seq   \n",
      "..                                                 ...   \n",
      "951              snRNA-seq of human retina - all cells   \n",
      "952                       Major cell cluster: Mesoderm   \n",
      "953                     Whole dataset: Raw counts only   \n",
      "954  Whole Taxonomy - MTG: Seattle Alzheimer's Dise...   \n",
      "955  Whole Taxonomy - DLPFC: Seattle Alzheimer's Di...   \n",
      "\n",
      "                             dataset_h5ad_path  dataset_total_cell_count  \n",
      "0    d7476ae2-e320-4703-8304-da5c42627e71.h5ad                       565  \n",
      "1    0895c838-e550-48a3-a777-dbcd35d30272.h5ad                       146  \n",
      "2    00ff600e-6e2e-4d76-846f-0eec4f0ae417.h5ad                       363  \n",
      "3    bdacc907-7c26-419f-8808-969eab3ca2e8.h5ad                      3799  \n",
      "4    a5d95a42-0137-496f-8a60-101e17f263c8.h5ad                      1324  \n",
      "..                                         ...                       ...  \n",
      "951  d6505c89-c43d-4c28-8c4f-7351a5fd5528.h5ad                   3177310  \n",
      "952  dcfa2614-7ca7-4d82-814c-350626eccb26.h5ad                   3267338  \n",
      "953  dcfd4feb-18a3-4b30-81d7-1b0c544a8ab3.h5ad                  11441407  \n",
      "954  c2876b1b-06d8-4d96-a56b-5304f815b99a.h5ad                   1226855  \n",
      "955  6f7fd0f1-a2ed-4ff1-80d3-33dde731cbc3.h5ad                   1309414  \n",
      "\n",
      "[956 rows x 11 columns]\n",
      "\n",
      "counts:\n",
      "\n",
      "columns: Index(['soma_joinid', 'organism', 'category', 'label', 'ontology_term_id',\n",
      "       'total_cell_count', 'unique_cell_count'],\n",
      "      dtype='object')\n",
      "\n",
      "      soma_joinid      organism        category            label  \\\n",
      "0               0  Homo sapiens             all               na   \n",
      "1               1  Homo sapiens           assay         Drop-seq   \n",
      "2               2  Homo sapiens           assay           inDrop   \n",
      "3               3  Homo sapiens           assay         MARS-seq   \n",
      "4               4  Homo sapiens           assay         Seq-Well   \n",
      "...           ...           ...             ...              ...   \n",
      "1896         1896  Mus musculus  tissue_general   exocrine gland   \n",
      "1897         1897  Mus musculus  tissue_general   prostate gland   \n",
      "1898         1898  Mus musculus  tissue_general  endocrine gland   \n",
      "1899         1899  Mus musculus  tissue_general      bone marrow   \n",
      "1900         1900  Mus musculus  tissue_general        optic cup   \n",
      "\n",
      "     ontology_term_id  total_cell_count  unique_cell_count  \n",
      "0                  na          86410648           51835017  \n",
      "1         EFO:0008722            338069             292631  \n",
      "2         EFO:0008780             51304              25652  \n",
      "3         EFO:0008796             70146              70146  \n",
      "4         EFO:0008919            255232             137955  \n",
      "...               ...               ...                ...  \n",
      "1896   UBERON:0002365             46731              15577  \n",
      "1897   UBERON:0002367            130135              37715  \n",
      "1898   UBERON:0002368             39966              13322  \n",
      "1899   UBERON:0002371            199699              90225  \n",
      "1900   UBERON:0003072               146                146  \n",
      "\n",
      "[1901 rows x 7 columns]\n",
      "\n",
      "var:\n",
      "\n",
      "columns: Index(['soma_joinid', 'feature_id', 'feature_name', 'feature_length', 'nnz',\n",
      "       'n_measured_obs'],\n",
      "      dtype='object')\n",
      "\n",
      "       soma_joinid          feature_id       feature_name  feature_length  \\\n",
      "0                0     ENSG00000237491          LINC01409            1059   \n",
      "1                1     ENSG00000188976              NOC2L            1244   \n",
      "2                2     ENSG00000187642              PERM1            2765   \n",
      "3                3     ENSG00000272512  ENSG00000272512.1            2086   \n",
      "4                4     ENSG00000188290               HES4             961   \n",
      "...            ...                 ...                ...             ...   \n",
      "52432        52432  ENSMUSG00002076766            Snord33              84   \n",
      "52433        52433  ENSMUSG00000118601            Gm53042             814   \n",
      "52434        52434  ENSMUSG00000118631            Gm53019            1593   \n",
      "52435        52435  ENSMUSG00000118645            Gm55062             480   \n",
      "52436        52436  ENSMUSG00000118652            Gm54771             358   \n",
      "\n",
      "            nnz  n_measured_obs  \n",
      "0       7266057        75339772  \n",
      "1      15080647        86077006  \n",
      "2        537346        75481456  \n",
      "3        805209        76397515  \n",
      "4      15402916        85816634  \n",
      "...         ...             ...  \n",
      "52432       612           73347  \n",
      "52433        10          301796  \n",
      "52434         5          301796  \n",
      "52435         0               0  \n",
      "52436         1          301796  \n",
      "\n",
      "[113875 rows x 6 columns]\n",
      "\n",
      "obs:\n",
      "\n",
      "columns: Index(['soma_joinid', 'dataset_id', 'assay', 'assay_ontology_term_id',\n",
      "       'cell_type', 'cell_type_ontology_term_id', 'development_stage',\n",
      "       'development_stage_ontology_term_id', 'disease',\n",
      "       'disease_ontology_term_id', 'donor_id', 'is_primary_data',\n",
      "       'observation_joinid', 'self_reported_ethnicity',\n",
      "       'self_reported_ethnicity_ontology_term_id', 'sex',\n",
      "       'sex_ontology_term_id', 'suspension_type', 'tissue',\n",
      "       'tissue_ontology_term_id', 'tissue_type', 'tissue_general',\n",
      "       'tissue_general_ontology_term_id', 'raw_sum', 'nnz', 'raw_mean_nnz',\n",
      "       'raw_variance_nnz', 'n_measured_vars', 'organism'],\n",
      "      dtype='object')\n",
      "\n",
      "         soma_joinid                            dataset_id      assay  \\\n",
      "0              69389  2f6a20f1-173d-4b8d-860b-c47ffea120fa  10x 3' v2   \n",
      "1              69390  2f6a20f1-173d-4b8d-860b-c47ffea120fa  10x 3' v2   \n",
      "2              69391  2f6a20f1-173d-4b8d-860b-c47ffea120fa  10x 3' v2   \n",
      "3              69392  2f6a20f1-173d-4b8d-860b-c47ffea120fa  10x 3' v2   \n",
      "4              69393  2f6a20f1-173d-4b8d-860b-c47ffea120fa  10x 3' v2   \n",
      "...              ...                                   ...        ...   \n",
      "3859649     12264935  d7291f04-fbbb-4d65-990a-f01fa44e915b  10x 3' v2   \n",
      "3859650     12264936  d7291f04-fbbb-4d65-990a-f01fa44e915b  10x 3' v2   \n",
      "3859651     12264937  d7291f04-fbbb-4d65-990a-f01fa44e915b  10x 3' v2   \n",
      "3859652     12264938  d7291f04-fbbb-4d65-990a-f01fa44e915b  10x 3' v2   \n",
      "3859653     12264939  d7291f04-fbbb-4d65-990a-f01fa44e915b  10x 3' v2   \n",
      "\n",
      "        assay_ontology_term_id  \\\n",
      "0                  EFO:0009899   \n",
      "1                  EFO:0009899   \n",
      "2                  EFO:0009899   \n",
      "3                  EFO:0009899   \n",
      "4                  EFO:0009899   \n",
      "...                        ...   \n",
      "3859649            EFO:0009899   \n",
      "3859650            EFO:0009899   \n",
      "3859651            EFO:0009899   \n",
      "3859652            EFO:0009899   \n",
      "3859653            EFO:0009899   \n",
      "\n",
      "                                                 cell_type  \\\n",
      "0                                   retina horizontal cell   \n",
      "1                                   retina horizontal cell   \n",
      "2                                   retina horizontal cell   \n",
      "3                                   retina horizontal cell   \n",
      "4                                   retina horizontal cell   \n",
      "...                                                    ...   \n",
      "3859649           sst chodl GABAergic cortical interneuron   \n",
      "3859650           sst chodl GABAergic cortical interneuron   \n",
      "3859651  L2/3-6 intratelencephalic projecting glutamate...   \n",
      "3859652                 sst GABAergic cortical interneuron   \n",
      "3859653  L2/3-6 intratelencephalic projecting glutamate...   \n",
      "\n",
      "        cell_type_ontology_term_id  development_stage  \\\n",
      "0                       CL:0000745  78-year-old stage   \n",
      "1                       CL:0000745  78-year-old stage   \n",
      "2                       CL:0000745  78-year-old stage   \n",
      "3                       CL:0000745  78-year-old stage   \n",
      "4                       CL:0000745  78-year-old stage   \n",
      "...                            ...                ...   \n",
      "3859649                 CL:4023121   8-week-old stage   \n",
      "3859650                 CL:4023121   8-week-old stage   \n",
      "3859651                 CL:4023040   8-week-old stage   \n",
      "3859652                 CL:4023017   7-week-old stage   \n",
      "3859653                 CL:4023040   8-week-old stage   \n",
      "\n",
      "        development_stage_ontology_term_id disease disease_ontology_term_id  \\\n",
      "0                           HsapDv:0000172  normal             PATO:0000461   \n",
      "1                           HsapDv:0000172  normal             PATO:0000461   \n",
      "2                           HsapDv:0000172  normal             PATO:0000461   \n",
      "3                           HsapDv:0000172  normal             PATO:0000461   \n",
      "4                           HsapDv:0000172  normal             PATO:0000461   \n",
      "...                                    ...     ...                      ...   \n",
      "3859649                     MmusDv:0000154  normal             PATO:0000461   \n",
      "3859650                     MmusDv:0000154  normal             PATO:0000461   \n",
      "3859651                     MmusDv:0000154  normal             PATO:0000461   \n",
      "3859652                     MmusDv:0000152  normal             PATO:0000461   \n",
      "3859653                     MmusDv:0000154  normal             PATO:0000461   \n",
      "\n",
      "         ... tissue_ontology_term_id  tissue_type tissue_general  \\\n",
      "0        ...          UBERON:0001786       tissue            eye   \n",
      "1        ...          UBERON:0001786       tissue            eye   \n",
      "2        ...          UBERON:0001786       tissue            eye   \n",
      "3        ...          UBERON:0001786       tissue            eye   \n",
      "4        ...          UBERON:0001786       tissue            eye   \n",
      "...      ...                     ...          ...            ...   \n",
      "3859649  ...          UBERON:0001384       tissue          brain   \n",
      "3859650  ...          UBERON:0001384       tissue          brain   \n",
      "3859651  ...          UBERON:0001384       tissue          brain   \n",
      "3859652  ...          UBERON:0022352       tissue          brain   \n",
      "3859653  ...          UBERON:0000411       tissue          brain   \n",
      "\n",
      "        tissue_general_ontology_term_id  raw_sum   nnz raw_mean_nnz  \\\n",
      "0                        UBERON:0000970   2157.0  1140     1.892105   \n",
      "1                        UBERON:0000970   1653.0   950     1.740000   \n",
      "2                        UBERON:0000970   1854.0   925     2.004324   \n",
      "3                        UBERON:0000970   1970.0  1059     1.860246   \n",
      "4                        UBERON:0000970   1966.0  1001     1.964036   \n",
      "...                                 ...      ...   ...          ...   \n",
      "3859649                  UBERON:0000955  14129.0  4737     2.982689   \n",
      "3859650                  UBERON:0000955  17296.0  5250     3.294476   \n",
      "3859651                  UBERON:0000955   6630.0  3316     1.999397   \n",
      "3859652                  UBERON:0000955   8906.0  3798     2.344918   \n",
      "3859653                  UBERON:0000955  17309.0  4668     3.708012   \n",
      "\n",
      "        raw_variance_nnz n_measured_vars      organism  \n",
      "0             122.226277           19884  homo_sapiens  \n",
      "1              76.797450           19884  homo_sapiens  \n",
      "2              73.692622           19884  homo_sapiens  \n",
      "3              72.729032           19884  homo_sapiens  \n",
      "4             127.932705           19884  homo_sapiens  \n",
      "...                  ...             ...           ...  \n",
      "3859649        40.448180           28102  mus_musculus  \n",
      "3859650        78.175412           28102  mus_musculus  \n",
      "3859651         8.890497           28102  mus_musculus  \n",
      "3859652        28.916028           28102  mus_musculus  \n",
      "3859653        88.272342           28102  mus_musculus  \n",
      "\n",
      "[28721975 rows x 29 columns]\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    datasets, counts, var, obs = get_metadata_and_datasets()\n",
    "except Exception as exc:\n",
    "    print_exception(exc)\n",
    "print(f\"datasets:\\n\\ncolumns: {datasets.columns}\\n\\n{datasets}\")\n",
    "print()\n",
    "print(f\"counts:\\n\\ncolumns: {counts.columns}\\n\\n{counts}\")\n",
    "print()\n",
    "print(f\"var:\\n\\ncolumns: {var.columns}\\n\\n{var}\")\n",
    "print()\n",
    "print(f\"obs:\\n\\ncolumns: {obs.columns}\\n\\n{obs}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3af73447-8c70-4d57-8ef7-f0d5db94ae33",
   "metadata": {},
   "source": [
    "We can merge some of the resulting DataFrames to create a summary\n",
    "DataFrame for export. As before, we will write the summary DataFrame\n",
    "to a `.parquet` file, so that later we can simply read the `.parquet`\n",
    "file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cd35fb05-71ba-4255-92b3-d9af0f893463",
   "metadata": {
    "results": "output",
    "session": "shared"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading summary parquet\n"
     ]
    }
   ],
   "source": [
    "summary_parquet = f\"{NLM_KN_DIR}/summary.parquet\"\n",
    "if not os.path.exists(summary_parquet):\n",
    "    # Define columns needed from obs\n",
    "    obs_columns = [\n",
    "        \"organism\",\n",
    "        \"tissue_general\",\n",
    "        \"tissue_general_ontology_term_id\",\n",
    "        \"assay\",\n",
    "        \"assay_ontology_term_id\",\n",
    "        \"dataset_id\",\n",
    "    ]\n",
    "\n",
    "    # Define columns required for summary\n",
    "    sum_columns = [\n",
    "        \"organism\",\n",
    "        \"tissue_general\",\n",
    "        \"tissue_general_ontology_term_id\",\n",
    "        \"collection_id\",\n",
    "        \"collection_name\",\n",
    "        \"collection_doi\",\n",
    "        \"assay\",\n",
    "        \"assay_ontology_term_id\",\n",
    "        \"dataset_id\",\n",
    "        \"dataset_title\",\n",
    "        \"dataset_h5ad_path\",\n",
    "    ]\n",
    "\n",
    "    print(\"Merging datasets and obs DataFrames\")\n",
    "    try:\n",
    "        summary = pd.merge(\n",
    "            datasets, obs[obs_columns].drop_duplicates(), on=\"dataset_id\"\n",
    "        )[sum_columns].drop_duplicates()\n",
    "    except Exception as exc:\n",
    "        print_exception(exc)\n",
    "\n",
    "    print(\"Writing summary parquet\")\n",
    "    summary.to_parquet(summary_parquet)\n",
    "\n",
    "    print(\"Writing summary CSV\")\n",
    "    summary_csv = f\"{NLM_KN_DIR}/summary.csv\"\n",
    "    summary.to_csv(summary_csv)\n",
    "\n",
    "else:\n",
    "    print(\"Reading summary parquet\")\n",
    "    summary = pd.read_parquet(summary_parquet)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5befbe72-994e-4a0f-8706-b69e1985376d",
   "metadata": {},
   "source": [
    "## Identification of publications corresponding to CELLxGENE datasets\n",
    "\n",
    "We notice that the datasets DataFrame contains a `citation` column,\n",
    "for example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "151b824a-1327-4c74-a220-f5bb6a99c941",
   "metadata": {
    "results": "output",
    "session": "shared"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Publication: https://doi.org/10.1038/s41467-022-29450-x Dataset Version: https://datasets.cellxgene.cziscience.com/f223d2b0-d878-443b-8bef-f182fe1f22d1.h5ad curated and distributed by CZ CELLxGENE Discover in Collection: https://cellxgene.cziscience.com/collections/bf325905-5e8e-42e3-933d-9a9053e9af80\n"
     ]
    }
   ],
   "source": [
    "print(datasets[\"citation\"].iloc[4])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "050da21e-9367-4f08-930f-5ddadd0db44c",
   "metadata": {},
   "source": [
    "The `citation` provides the DOI, but not the title of the\n",
    "publication. Note that we will need the title later in Chapter 02:\n",
    "E-Utilities. So, we examine the `collection_name` and `dataset_title`\n",
    "columns:\n",
    "\n",
    "See: [Chapter-02-E-Utilities.ipynb](Chapter-02-E-Utilities.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "90807918-6ad6-49f4-8474-4af6b2e36bf9",
   "metadata": {
    "results": "output",
    "session": "shared"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "collection_name    Single-cell Atlas of common variable immunodef...\n",
      "dataset_title                       Steady-state B cells - scRNA-seq\n",
      "Name: 4, dtype: object\n"
     ]
    }
   ],
   "source": [
    "print(datasets[[\"collection_name\", \"dataset_title\"]].iloc[4, :])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9351106f-9b03-4e12-a18e-f300bea22388",
   "metadata": {},
   "source": [
    "But it appears we still need to find the title by some method. So, we\n",
    "write a function that requests the DOI, then parses the resulting\n",
    "page, most likely from the publisher, to find the title."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "efdf4711-18f4-4951-8b98-5495fd01c9bd",
   "metadata": {
    "results": "silent",
    "session": "shared",
    "tangle": "../py/CELLxGENE.py"
   },
   "outputs": [],
   "source": [
    "def get_title(citation):\n",
    "    \"\"\"Get the title given a dataset citation. Note that only wget\n",
    "    succeeded for Cell Press journals, and neither requests nor wget\n",
    "    succeeded for The EMBO Journal and Science.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    citation : str\n",
    "        Dataset citation\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    title : str\n",
    "        Title of publication associated with the dataset\n",
    "    \"\"\"\n",
    "    # Need a default return value\n",
    "    title = None\n",
    "\n",
    "    # Compile patterns for finding the publication URL and article\n",
    "    # title\n",
    "    p1 = re.compile(\"Publication: (.*) Dataset Version:\")\n",
    "    p2 = re.compile(\"articleName : '(.*)',\")\n",
    "\n",
    "    # Assign CSS selectors for selecting article title elements\n",
    "    selectors = [\n",
    "        \"h1.c-article-title\",\n",
    "        \"h1.article-header__title.smaller\",\n",
    "        \"div.core-container h1\",\n",
    "        \"h1.content-header__title.content-header__title--xx-long\",\n",
    "        \"h1#page-title.highwire-cite-title\",\n",
    "    ]\n",
    "\n",
    "    # Find the publication URL\n",
    "    m1 = p1.search(citation)\n",
    "    if not m1:\n",
    "        logging.warning(f\"Could not find citation URL for {citation}\")\n",
    "        return title\n",
    "    citation_url = m1.group(1)\n",
    "    print(f\"Getting title for citation URL: {citation_url}\")\n",
    "\n",
    "    # Attempt to get the publication page using requests\n",
    "    print(f\"Trying requests\")\n",
    "    sleep(HTTPS_SLEEP)\n",
    "    response = requests.get(citation_url)\n",
    "    try_wget = True\n",
    "    if response.status_code == 200:\n",
    "        html_data = response.text\n",
    "\n",
    "        # Got the page, so parse it, and try each selector\n",
    "        fullsoup = BeautifulSoup(html_data, features=\"lxml\")\n",
    "        for selector in selectors:\n",
    "            selected = fullsoup.select(selector)\n",
    "            if selected:\n",
    "\n",
    "                # Selected the article title, so assign it\n",
    "                if len(selected) > 1:\n",
    "                    logging.warning(\n",
    "                        f\"Selected more than one element using {selector} on soup from {citation_url}\"\n",
    "                    )\n",
    "                title = selected[0].text\n",
    "                try_wget = False\n",
    "                break\n",
    "\n",
    "    if try_wget:\n",
    "\n",
    "        # Attempt to get the publication page using wget\n",
    "        print(f\"Trying wget\")\n",
    "        sleep(HTTPS_SLEEP)\n",
    "        completed_process = subprocess.run(\n",
    "            [\"curl\", \"-L\", citation_url], capture_output=True\n",
    "        )\n",
    "        html_data = completed_process.stdout\n",
    "\n",
    "        # Got the page, so parse it, and search for the title\n",
    "        fullsoup = BeautifulSoup(html_data, features=\"lxml\")\n",
    "        found = fullsoup.find_all(\"script\")\n",
    "        if found and len(found) > 4:\n",
    "            m2 = p2.search(found[4].text)\n",
    "            if m2:\n",
    "                title = m2.group(1)\n",
    "\n",
    "    print(f\"Found title: '{title}' for citation URL: {citation_url}\")\n",
    "\n",
    "    return title\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e41f09d-b388-4849-856b-910f266677c2",
   "metadata": {},
   "source": [
    "Next we call the function for an example citation (again using\n",
    "exception handling since accessing an external resource):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9745043a-7876-448d-9747-f3a8eebdbd66",
   "metadata": {
    "results": "output",
    "session": "shared"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Getting title for citation URL: https://doi.org/10.1038/s41467-022-29450-x\n",
      "Trying requests\n",
      "Found title: 'Single-cell Atlas of common variable immunodeficiency shows germinal center-associated epigenetic dysregulation in B-cell responses' for citation URL: https://doi.org/10.1038/s41467-022-29450-x\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    citation = datasets[\"citation\"].iloc[4]\n",
    "    title = get_title(citation)\n",
    "except Exception as exc:\n",
    "    print_exception(exc)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb5cf5d3-1c01-438d-b0de-7c7ed9c60925",
   "metadata": {},
   "source": [
    "Note that the function attempts to use `requests`, and if it fails,\n",
    "`wget`, since some publishers respond to one, but not the other. The\n",
    "selectors were discovered by manually inspecting the pages returned\n",
    "for the human lung cell datasets using Google Chrome Developer Tools.\n",
    "\n",
    "## Determine the dataset filename and download the dataset file.\n",
    "\n",
    "Following a notebook found in a CZI repository, we write a function to\n",
    "find the dataset filename, and to download the dataset file, given a\n",
    "row of the datasets DataFrame obtained above:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "53a7462c-af46-4ad7-a294-e6b46ec7473d",
   "metadata": {
    "results": "silent",
    "session": "shared",
    "tangle": "../py/CELLxGENE.py"
   },
   "outputs": [],
   "source": [
    "def get_and_download_dataset_h5ad_file(dataset_series):\n",
    "    \"\"\"Get the dataset filename and download the dataset file.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    dataset_series : pd.Series\n",
    "        A row from the dataset DataFrame\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    dataset : str\n",
    "       The dataset filename\n",
    "    \"\"\"\n",
    "    # Need a default return value\n",
    "    dataset_filename = None\n",
    "\n",
    "    # Get the dataset object\n",
    "    collection_id = dataset_series.collection_id\n",
    "    dataset_id = dataset_series.dataset_id\n",
    "    dataset_url = f\"{CELLXGENE_API_URL_BASE}/curation/v1/collections/{collection_id}/datasets/{dataset_id}\"\n",
    "    sleep(HTTPS_SLEEP)\n",
    "    response = requests.get(dataset_url)\n",
    "    response.raise_for_status()\n",
    "    if response.status_code != 200:\n",
    "        logging.error(f\"Could not get dataset for id {dataset_id}\")\n",
    "        return\n",
    "\n",
    "    data = response.json()\n",
    "    if dataset_id != data[\"dataset_id\"]:\n",
    "        logging.error(\n",
    "            f\"Response dataset id: {data['dataset_id']} does not equal specified dataset id: {dataset_id}\"\n",
    "        )\n",
    "        return\n",
    "\n",
    "    # Find H5AD files, if possible\n",
    "    assets = data[\"assets\"]\n",
    "    for asset in assets:\n",
    "        if asset[\"filetype\"] != \"H5AD\":\n",
    "            continue\n",
    "\n",
    "        # Found an H5AD file, so download it, if needed\n",
    "        dataset_filename = f\"{dataset_id}.{asset['filetype']}\"\n",
    "        dataset_filepath = f\"{CELLXGENE_DIR}/{dataset_filename}\"\n",
    "        if not os.path.exists(dataset_filepath):\n",
    "            print(f\"Downloading dataset file: {dataset_filepath}\")\n",
    "            with requests.get(asset[\"url\"], stream=True) as response:\n",
    "                response.raise_for_status()\n",
    "                with open(dataset_filepath, \"wb\") as df:\n",
    "                    for chunk in response.iter_content(chunk_size=1024 * 1024):\n",
    "                        df.write(chunk)\n",
    "            print(f\"Dataset file: {dataset_filepath} downloaded\")\n",
    "\n",
    "        else:\n",
    "            print(f\"Dataset file: {dataset_filepath} exists\")\n",
    "\n",
    "    return dataset_filename\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00b99028-85bd-484f-849b-fd63778172df",
   "metadata": {},
   "source": [
    "Then call it using the first row of the datasets DataFrame obtained\n",
    "above, and print the result (we'll use exception handling when\n",
    "accessing an external resource from now on):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "78acd30f-629d-470f-adfe-cb2e2bab5acb",
   "metadata": {
    "results": "output",
    "session": "shared"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dataset file: ../data/cellxgene/a5d95a42-0137-496f-8a60-101e17f263c8.H5AD exists\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    dataset_series = datasets.iloc[4]\n",
    "    get_and_download_dataset_h5ad_file(dataset_series)\n",
    "except Exception as exc:\n",
    "    print_exception(exc)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a611e8f8-2648-41e3-b5f3-3c05edc05100",
   "metadata": {},
   "source": [
    "Next, in Chapter 02 we write functions to search PubMed for the title\n",
    "and identifiers.\n",
    "\n",
    "See: [Chapter-02-E-Utilities.ipynb](Chapter-02-E-Utilities.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11c03e5a-0b55-4e6f-a0e7-1f622f58550a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08310b22-e1c9-42bd-925e-261af38812e3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
